q2_moshpit/kraken2/classification.py

```diff
@@ -79,25 +81,149 @@ def _classify_kraken(
+    )
+    bracken_output_fp = os.path.join(
+        tmp_dir, f"{sample_id}.bracken.output.txt"
+    )
+    bracken_report_fp = os.path.join(
+        tmp_dir, f"{sample_id}.bracken.report.txt"
+    )
+    cmd = [
+        "bracken",
+        "-d", bracken_db,
+        "-i", kraken2_report_fp,
+        "-o", bracken_output_fp,
+        "-w", bracken_report_fp,
+        "-t", str(threshold),
+        "-r", str(read_len),
+        "-l", level,
+    ]
+    try:
+        run_command(cmd=cmd, verbose=True)
+    except subprocess.CalledProcessError as e:
+        # TODO: what should be the behaviour when no reads could be classified?
+        raise Exception(
+            "An error was encountered while running Bracken, "
+            f"(return code {e.returncode}), please inspect "
+            "stdout and stderr to learn more."
+        )
+    bracken_table = pd.read_csv(bracken_output_fp, sep="\t", index_col=0)
+    bracken_table["sample_id"] = sample_id
+
+    return bracken_table
+
+
+def _estimate_bracken(
+        kraken_reports: Kraken2ReportDirectoryFormat,
+        bracken_db: BrackenDBDirectoryFormat,
+        threshold: int,
+        read_len: int,
+        level: str
+) -> pd.DataFrame:
+    bracken_tables = []
+    with tempfile.TemporaryDirectory() as tmpdir:
+        try:
+            for report_dir in kraken_reports.path.iterdir():
+                bracken_table = _run_bracken_one_sample(
+                    bracken_db=str(bracken_db),
+                    kraken2_report_dir=report_dir,
+                    tmp_dir=tmpdir, threshold=threshold,
+                    read_len=read_len, level=level
+                )
+                bracken_tables.append(bracken_table)
+        except subprocess.CalledProcessError as e:
+            raise Exception(
+                "An error was encountered while running Bracken, "
+                f"(return code {e.returncode}), please inspect "
+                "stdout and stderr to learn more."
+            )
+
+    bracken_table = pd.concat(bracken_tables).reset_index()
+    bracken_table = bracken_table.pivot(
+        index="sample_id", columns="name", values="new_est_reads"
+    )
+
+    return bracken_table
+
+
+def _get_available_lens(bracken_db: BrackenDBDirectoryFormat) -> List[int]:
+    pattern = r'.+database(\d{2,})mers\.kmer_distrib$'
+    lengths = []
+    for db in bracken_db.path.iterdir():
+        lengths.extend(re.findall(pattern, str(db)))
+    return sorted([int(x) for x in lengths])
+
+
+def _is_reads(seqs):
+    return isinstance(
+            seqs, (SingleLanePerSamplePairedEndFastqDirFmt,
+                   SingleLanePerSampleSingleEndFastqDirFmt)
+    )
+
+
 def classify_kraken(
     seqs: Union[
         SingleLanePerSamplePairedEndFastqDirFmt,
         SingleLanePerSampleSingleEndFastqDirFmt,
         MultiMAGSequencesDirFmt,
     ],
-    db: Kraken2DBDirectoryFormat,
+    kraken2_db: Kraken2DBDirectoryFormat,
+    bracken_db: BrackenDBDirectoryFormat,
```

Comment:
Would it make sense for kraken and braken to be two separate actions, and then combined in a `Pipeline`?

Reply:
Having thought about it a bit more, I agree: it would solve our issue of outputting a "random"/empty FeatureTable when MAGs were used as input. What we could do instead is:
- have a `classify-kraken` action which can accept **reads/MAGs** and output the Kraken reports/outputs
- have a `reestimate-bracken` pipeline which can **only** accept **reads**, use the previous action to run Kraken and then add the Bracken part on top - this would then output the feature table (we probably should also output the original Kraken reports?)

What do you think?

Reply:
Hey @gregcaporaso, FYI: in my last commit I split the existing action into two separate ones (as described above). The pipeline is called `classify-kraken-bracken` for now - I'll be more than happy to hear other name suggestions 😅

Reply:
That sounds great! I haven't had a chance to jump into testing this yet unfortunately, but I'm expecting to get back to it around the beginning of next week. Sorry for the delay!